File: tools/openclinxr/asset-pipeline/makeclothes/stamp-garment-provenance.ts
/**
 * #596 — stamp garment mesh extras with the source `.mhclo` the bake fitted from.
 *
 * The materializer holds the path at fit time (`read_hair_mhclo_licence`, `Mhclo.load`)
 * and historically discarded it. Without mesh extras the only signal is the material
 * name — which is how a peds_upper shell shipped as `hospital_gown`. This stamp is the
 * durable record clause (1) of `the-patient-gown-is-a-gown-class-asset` reads.
 *
 * claimScope: mesh.extras { sourceMhclo, garmentClass, licence } on MakeClothes /
 * real_garment meshes.
 * notEvidenceFor: appearance, fit quality, Anny retirement.
 */

import { readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { NodeIO } from "@gltf-transform/core";

export type GarmentProvenanceStamp = {
  sourceMhclo: string;
  garmentClass: string;
  licence: string;
};

/**
 * Mesh-name substring → provenance. Longest match wins. Hair/eyes/brows/teeth are
 * excluded by the planted contract's filter; footwear and clothing are stamped.
 */
export const GARMENT_PROVENANCE_BY_MESH_SUBSTR: ReadonlyArray<{
  match: RegExp;
  stamp: GarmentProvenanceStamp;
}> = [
  {
    match: /lab_coat|labcoat|crudelabcoatopen/i,
    stamp: {
      sourceMhclo: "crudelabcoatopen.mhclo",
      garmentClass: "labcoat",
      licence: "CC0",
    },
  },
  {
    match: /scrub_shirt/i,
    stamp: {
      sourceMhclo: "Scrub_Shirt.mhclo",
      garmentClass: "scrub",
      licence: "CC-BY",
    },
  },
  {
    match: /scrub_pants/i,
    stamp: {
      sourceMhclo: "Scrub_Pants.mhclo",
      garmentClass: "scrub",
      licence: "CC-BY",
    },
  },
  {
    match: /toigo_t_shirt|toigo_basic_tucked_t_shirt|toigo_basic_tucked_t-shirt/i,
    stamp: {
      sourceMhclo: "toigo_basic_tucked_t-shirt.mhclo",
      garmentClass: "street",
      licence: "CC0",
    },
  },
  {
    match: /cargo_pants/i,
    stamp: {
      sourceMhclo: "cargo_pants.mhclo",
      garmentClass: "street",
      licence: "CC0",
    },
  },
  {
    match: /footwear_toigo_flats|toigo_flats/i,
    stamp: {
      sourceMhclo: "toigo_flats.mhclo",
      garmentClass: "footwear",
      licence: "CC0",
    },
  },
  {
    match: /footwear.*mj_cloth|toigo_mj_cloth/i,
    stamp: {
      sourceMhclo: "toigo_mj_cloth_shoes.mhclo",
      garmentClass: "footwear",
      licence: "CC0",
    },
  },
  {
    match: /culturalibre_male_boots|male_boots/i,
    stamp: {
      sourceMhclo: "culturalibre_male_boots.mhclo",
      garmentClass: "footwear",
      licence: "CC-0",
    },
  },
];

const SKIP = /eyes|hair|eyelash|eyebrow|teeth|tongue|declared_upper/i;

export function provenanceForMeshName(name: string): GarmentProvenanceStamp | null {
  if (!/real_garment|makeclothes_library/i.test(name)) return null;
  if (SKIP.test(name)) return null;
  for (const row of GARMENT_PROVENANCE_BY_MESH_SUBSTR) {
    if (row.match.test(name)) return row.stamp;
  }
  return null;
}

export async function stampGarmentProvenanceOnGlb(
  glbPath: string,
  overrides: Readonly<Record<string, GarmentProvenanceStamp>> = {},
): Promise<{ stamped: string[]; skipped: string[] }> {
  const io = new NodeIO();
  const doc = await io.read(glbPath);
  const stamped: string[] = [];
  const skipped: string[] = [];
  for (const mesh of doc.getRoot().listMeshes()) {
    const name = mesh.getName() ?? "";
    const fromOverride = overrides[name];
    const stamp = fromOverride ?? provenanceForMeshName(name);
    if (!stamp) {
      if (/real_garment|makeclothes_library/i.test(name) && !SKIP.test(name)) {
        skipped.push(name);
      }
      continue;
    }
    mesh.setExtras({
      ...(mesh.getExtras() ?? {}),
      sourceMhclo: stamp.sourceMhclo,
      garmentClass: stamp.garmentClass,
      licence: stamp.licence,
    });
    stamped.push(name);
  }
  await io.write(glbPath, doc);
  return { stamped, skipped };
}

/** CLI: `pnpm exec tsx …/stamp-garment-provenance.ts <glb> [<glb>…]` */
async function main(argv: string[]): Promise<void> {
  const paths = argv.filter((a) => a.endsWith(".glb"));
  if (paths.length === 0) {
    console.error("usage: stamp-garment-provenance.ts <glb> [<glb>…]");
    process.exit(2);
  }
  for (const p of paths) {
    const abs = path.resolve(p);
    const result = await stampGarmentProvenanceOnGlb(abs);
    console.log(
      JSON.stringify({
        glb: abs,
        stamped: result.stamped,
        skipped: result.skipped,
      }),
    );
    // Touch mtime proof for callers that key off write.
    writeFileSync(abs, readFileSync(abs));
  }
}

const isCli =
  typeof process !== "undefined" &&
  process.argv[1] != null &&
  /stamp-garment-provenance\.(ts|js|mjs|cjs)$/.test(process.argv[1]);
if (isCli) {
  main(process.argv.slice(2)).catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

File: tools/openclinxr/asset-pipeline/makeclothes/stamp_garment_provenance.py
"""#596 — stamp mesh.extras {sourceMhclo, garmentClass, licence} into a shipped GLB.

Python twin of stamp-garment-provenance.ts so Blender bakes can persist the path they
already hold at fit time without a Node round-trip. JSON-chunk rewrite only; BIN verbatim.
"""

from __future__ import annotations

import json
import re
import struct
from pathlib import Path
from typing import Iterable

# Longest-match order mirrors stamp-garment-provenance.ts
_RULES: list[tuple[re.Pattern[str], dict[str, str]]] = [
    (re.compile(r"lab_coat|labcoat|crudelabcoatopen", re.I), {
        "sourceMhclo": "crudelabcoatopen.mhclo",
        "garmentClass": "labcoat",
        "licence": "CC0",
    }),
    (re.compile(r"scrub_shirt", re.I), {
        "sourceMhclo": "Scrub_Shirt.mhclo",
        "garmentClass": "scrub",
        "licence": "CC-BY",
    }),
    (re.compile(r"scrub_pants", re.I), {
        "sourceMhclo": "Scrub_Pants.mhclo",
        "garmentClass": "scrub",
        "licence": "CC-BY",
    }),
    (re.compile(r"toigo_t_shirt|toigo_basic_tucked_t_shirt|toigo_basic_tucked_t-shirt", re.I), {
        "sourceMhclo": "toigo_basic_tucked_t-shirt.mhclo",
        "garmentClass": "street",
        "licence": "CC0",
    }),
    (re.compile(r"cargo_pants", re.I), {
        "sourceMhclo": "cargo_pants.mhclo",
        "garmentClass": "street",
        "licence": "CC0",
    }),
    (re.compile(r"footwear_toigo_flats|toigo_flats", re.I), {
        "sourceMhclo": "toigo_flats.mhclo",
        "garmentClass": "footwear",
        "licence": "CC0",
    }),
]

_SKIP = re.compile(r"eyes|hair|eyelash|eyebrow|teeth|tongue|declared_upper", re.I)
_GARMENT = re.compile(r"real_garment|makeclothes_library", re.I)


def provenance_for_mesh_name(name: str) -> dict[str, str] | None:
    if not _GARMENT.search(name) or _SKIP.search(name):
        return None
    for pat, stamp in _RULES:
        if pat.search(name):
            return dict(stamp)
    return None


def stamp_garment_provenance_glb(
    path: str | Path,
    overrides: dict[str, dict[str, str]] | None = None,
) -> list[str]:
    """Write provenance extras onto matching meshes. Returns stamped mesh names."""
    path = Path(path)
    overrides = overrides or {}
    data = bytearray(path.read_bytes())
    if data[:4] != b"glTF":
        raise RuntimeError(f"#596: not a GLB: {path}")
    json_len = struct.unpack("<I", data[12:16])[0]
    json_end = 20 + json_len
    gltf = json.loads(data[20:json_end])
    stamped: list[str] = []
    for mesh in gltf.get("meshes", []):
        name = mesh.get("name") or ""
        stamp = overrides.get(name) or provenance_for_mesh_name(name)
        if not stamp:
            continue
        extras = mesh.get("extras") or {}
        if not isinstance(extras, dict):
            extras = {}
        extras.update(stamp)
        mesh["extras"] = extras
        stamped.append(name)
    new_json = json.dumps(gltf, separators=(",", ":"))
    new_json += " " * ((4 - len(new_json) % 4) % 4)
    bin_chunk = data[json_end:]
    out = bytearray()
    out += b"glTF"
    out += struct.pack("<II", 2, 12 + 8 + len(new_json) + len(bin_chunk))
    out += struct.pack("<I", len(new_json)) + b"JSON"
    out += new_json.encode("utf-8")
    out += bin_chunk
    path.write_bytes(out)
    print(f"GLB_GARMENT_PROVENANCE_STAMP {path} meshes {','.join(stamped)}")
    return stamped


def stamp_many(paths: Iterable[str | Path]) -> None:
    for p in paths:
        stamp_garment_provenance_glb(p)


if __name__ == "__main__":
    import sys

    if len(sys.argv) < 2:
        raise SystemExit("usage: stamp_garment_provenance.py <glb> [<glb>…]")
    stamp_many(sys.argv[1:])
